rfcs/2022-04-20-12187-log-namespacing.md

```diff
@@ -0,0 +1,422 @@
+keys will be used. Previously this was useful so users could choose names that didn't collide with other data in the event.
+With namespacing, this is no longer a concern.
+
+Similar to the Global Log Schema, some sources allow choosing key names where some data will be stored. This will also be
+removed when the "Vector" namespace is being used, in favor of static names (usually what the default was). An example
+is the "key_field" and "headers_key" in the kafka source.
+
+Many transforms / sinks rely on the "Global Log Schema" to get/modify info such as the timestamp. Since the log
+schema should be customizable per source, that means the transforms and sinks need to know which log namespace
+is being used by each log. In order to accomplish this, the `log_namespace` event metadata field will be set to
+`vector` if the new namespace is being used.
+
+
+
+### Types of Data
+
+#### Data
+
+This is the main content of a log.  It will be placed at the root of the event.
+What is considered the "main content" here is not very well-defined, but should generally be what most people would expect
+to receive from the source. It is not necessarily the "log message" itself.
+
+A good example is the "Datadog Agent" source. The root is the fields received by the Datadog agent. The log content is
+nested.
+
+Others such as the "socket" source will have the log content directly at the root.
+
+There is a special case when the codec is "bytes", since the data is just a string in that case. That means
+that the root is a string. Historically this has been forbidden, but it is possible to allow this.
+An example of this is the "socket" source, with the "bytes" codec.
+
+
+#### Metadata
+
+This is any useful data that is not placed on the event. This will be stored in event metadata. Some improvements will be needed to
+event metadata to make this easier to use, as described below. This currently contains fields such as `datadog_api_key`.
+Other "vector" metadata such as `ingest_timestamp` and `source_type` will be added here, as well as additional metadata from
+sources, such as `key` and `headers` from the `kafka` source.
+
+### Codecs
+
+#### Bytes
+
+With the "Legacy" namespace, this codec decodes the input as-is, and places it under a "message" key (configurable with global log schema).
+With the "Vector" namespace, this will be placed under the "data" key. If there is no additional metadata stored
+on the event, then it will be placed on the root instead. The placement will be decided for each source independently.
+
+#### Json
+
+With the "Legacy" namespace, this codec decodes the input as-is, and places it under a "message" key (configurable with global log schema).
+With the "Vector" namespace, this will be placed under the "data" key. If there is no additional metadata stored
+on the event, then it will be placed on the root instead. The placement will be decided for each source independently.
```

Comment:
I do wonder if this will hurt UX.

Say I have two sources, both configured with the `json` decoder, one source does have metadata, the other doesn't, now one has the json stored in `data`, the other in the root. If I point both sources to a `remap` transform, and I want to delete the `foo` field that is part of the JSON payload of both sources, I now have to either remove both `.foo` and `.data.foo`, or conditionally check which one exists (but this one runs the risk of the data set to the root itself having a nested `.data` field), or have to add a new transform to move all decoded data on the root back into `data` for that one source.